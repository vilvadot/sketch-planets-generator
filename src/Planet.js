import { Random } from "./Random";
import { Moon } from "./Moon";

  // ------------ Seed --------------
  //  1      2        3         4         5-10
  //  size   moons    period    type      ??? 

export class Planet {
  constructor(seed) {
    this.seed = seed.toString();
    this.size = this._generateSize();
    this.moons = this._generateMoons();
    this.period = this._generatePeriod();
    this.type = this._generateType();
  }

  _generateType() {
    const TYPES = {
      Gas: 'Gas',
      Terrestrial: 'Terrestrial',
      Ocean: 'Ocean',
      Rocky: 'Rocky',
      Ice: 'Ice',
    }
    const possibleValues = [TYPES.Gas, TYPES.Gas, TYPES.Terrestrial, TYPES.Ocean, TYPES.Rocky, TYPES.Ice, TYPES.Gas, TYPES.Gas, TYPES.Rocky, TYPES.Ice]
    
    const seed = this.seed[3];

    return possibleValues[seed]
  }

  _generateSize() {
    const seed = this.seed[0];

    const lastDigit = seed.toString()[seed.toString().length - 1];
    return Number(lastDigit) * 2;
  }

  _generateMoons() {
    const moons = []
    const generator = new Random(this.seed[1] * 2);

    const firstDigit = generator.next().toString()[0];
    const numberOfMoons = Number(firstDigit)

    times(() => {
      moons.push(new Moon(generator.next(), this.size))
    }, numberOfMoons)

    return moons
  }

  _generatePeriod() {
    return this.seed[2] * 1.5 + 1;
  }
}

const times = (callback, numberOfTimes) => {
  for(let i = 0; i <= numberOfTimes; i++ ){
    callback()
  }
}
