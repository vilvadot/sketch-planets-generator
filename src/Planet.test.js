import { Planet } from "./Planet";
import { Random } from "./Random";

describe("Planet", () => {
  it("can be recreated from a seed", () => {
    const seed = new Random(12345678910).next();

    const planet = new Planet(seed);

    expect(planet).toMatchSnapshot();
  })
});
