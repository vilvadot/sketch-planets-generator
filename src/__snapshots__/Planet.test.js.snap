// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Planet can be recreated from a seed 1`] = `
Planet {
  "moons": Array [
    Moon {
      "distance": 23.760000000000005,
      "orbitalPeriod": 12,
      "parentSize": 2,
      "seed": "224636690",
      "size": 1.6,
    },
    Moon {
      "distance": 15.400000000000002,
      "orbitalPeriod": 46,
      "parentSize": 2,
      "seed": "192597404",
      "size": 0.8,
    },
    Moon {
      "distance": 58.52000000000001,
      "orbitalPeriod": 13,
      "parentSize": 2,
      "seed": "726712999",
      "size": 5.6000000000000005,
    },
  ],
  "period": 1,
  "seed": "1807983583",
  "size": 2,
  "type": "Gas",
}
`;
